### DISCUSSION TITLE
The docs page layout is abnormal

### DISCUSSION THREAD

**Discussion by fz-gaojian** [fastapi/fastapi#6648] [Questions] [answered]
Local service running with fastapi

visit docs page, the docs page layout is abnormal.

**Comment by leo-lee-4231** [comment]
I meet the same problem. The css of docs page seems not correct. It shows like this picture.

![image](https://user-images.githubusercontent.com/26795087/118911679-2d68c180-b959-11eb-8d2a-0572bbe9e62b.png)

It seems because it add a button element to contains the origin content, after I check the difference of the correct doc pages.

![image](https://user-images.githubusercontent.com/26795087/118914311-93efde80-b95d-11eb-9821-36151f1db2d4.png)

**Comment by boh5** [comment]
![fastapi-swagger](https://user-images.githubusercontent.com/12136712/118913058-5722e800-b95b-11eb-9883-063e1b96c76f.png)
same problem! occured about 1 hour ago. before that, every thing is ok

**Comment by ArcLightSlavik** [comment]
swagger-ui-dist had a release 5 hours ago (3.49.0) which probably broke this, could you give the example endpoint where this issue comes up?

After this you can do:

```
from fastapi import applications
from fastapi.openapi.docs import get_swagger_ui_html


def swagger_monkey_patch(*args, **kwargs):
    """
    Wrap the function which is generating the HTML for the /docs endpoint and 
    overwrite the default values for the swagger js and css.
    """
    return get_swagger_ui_html(
        *args, **kwargs,
        swagger_js_url="https://cdn.jsdelivr.net/npm/swagger-ui-dist@3.48.0/swagger-ui-bundle.js",
        swagger_css_url="https://cdn.jsdelivr.net/npm/swagger-ui-dist@3.48.0/swagger-ui.css"
    )


# Actual monkey patch
applications.get_swagger_ui_html = swagger_monkey_patch


# Your normal code ...
app = FastAPI()
```

**Comment by grillazz** [comment]
@ArcLightSlavik monkey patch tested and working ok

**Comment by tiangolo** [comment]
Thanks for the help @ArcLightSlavik ! :bow: 

@fz-gaojian, @leo-lee-4231, @dilless, or anyone else, could you please provide a simple example app that replicates the problem?

I have tried several things and I can't replicate it locally yet, so I can't check what would solve the problem, what version of Swagger UI to pin, etc.

If I could replicate the problem, I would be able to release a quick version of FastAPI with the temporary pin to fix it. And probably also report it to Swagger UI's repo, as it seems it hasn't been reported yet.

**Comment by fz-gaojian** [comment]
@tiangolo 
Sorry! due to information security, I can only describe.

fastapi==0.65.1
uvicorn[standard]==0.13.0

After starting the service and go to visit /docs will reappear.

If there is no reappear, you can try to clear the browser cache.

**Comment by dvarrazzo** [comment]
Probably swagger pushed a broken resource on their CDN, now they are fixing it. This looked broken until minutes ago, the bottom window was hard-refreshed with a Ctrl-F5
![image](https://user-images.githubusercontent.com/199429/118955822-58eeaa80-b95f-11eb-925d-2eff7b4fc877.png)

---
edit: at 2021-05-20T09:43:30+00:00 still hit and miss

**Comment by kozhushman** [comment]
Same here, still broken.
Monkey patch is working fine, old bundles (from https://pypi.org/project/fastapi-offline/ for example) also doing nice.
This is not FastAPI problem, but Swagger's
I suggest closing this issue - fixes are present here, problem not on FastAPI side, so discussion have no point.
But In previous similar cases @tiangolo was pinning swagger version, so may be similar solution here

**Comment by ArcLightSlavik** [comment]
Closing this issue would mean other people will have a hard time finding it, because no one checks closed issues.

**Comment by dvarrazzo** [comment]
I think if fastapi can do something is to bundle swagger static files, serve that by default, and provide an option to use the CDN instead.

Looking at the API docs is not something so resource-intensive to require a CDN, for most of the projects.

**Comment by tiangolo** [comment]
Yeah, something around bundling a local version is on my plans, but those assets require an independent build step to include them, they can't/shouldn't be added to GitHub compiled. And it would require a new version of FastAPI every time they release a new version, so people would be stuck with old versions of Swagger and ReDoc even if new versions fixed bugs and added features, and it would increase the size of the distributable (but maybe that's not that significant?).

Anyway, that's something I have planned to explore how to solve in the future.

---

Now, for this specific problem, I would release a version pinning the version that worked if I was able to replicate the problem and be sure myself what's the error and what fixes it. But up to now, I haven't been able to replicate it once, and sadly no one has reported an example app that shows the problem. :pensive:

And if it's not related to the specific app, I don't understand why I haven't been able to replicate the error myself with all the examples I've tried. It seems it could be a regional cache issue. :thinking:

**Comment by ArcLightSlavik** [comment]
I don't think the size of the distribution matters, there was some ai python library that is 800mb. 
I also don't think people being stuck on an old version of Swagger or Redoc matters that much, for me it is the same idea as requirements.txt we already have, locking this would solve the issue we have right now (similar issues with docs happened before).

One of the solutions I imagine is a separate package with all the CDN stuff, which we upload on pypi and have the same version as the npm version. This would mean we could have ranges when people download FastAPI.

Basically what I am saying, locking the version and doing offline download is a good idea.

I do think that the discussion should probably be moved to a separate ticket.

**Comment by andristeiner** [comment]
> sadly no one has reported an example app that shows the problem. 
> 
> And if it's not related to the specific app, I don't understand why I haven't been able to replicate the error myself with all the examples I've tried. It seems it could be a regional cache issue.

It looks like it. I'm living in Switzerland, the CSS is served from Frankfurt (Germany) according to the headers, and i can replicate the problem even with the "simplest FastAPI file" from [First Steps](https://fastapi.tiangolo.com/tutorial/first-steps/#first-steps):

```
from fastapi import FastAPI

app = FastAPI()

@app.get("/")
async def root():
    return {"message": "Hello World"}
```

does led to:

![image](https://user-images.githubusercontent.com/249843/118986296-84818d00-b97f-11eb-9995-5c949e12b9df.png)

At the moment, i work around this problem by adding `width: 100%` to `.swagger-ui button` through my browsers dev tools. It's not perfect and it does not survive a reload, but i'll do for me until Swagger can come up with a permanent fix.

**Comment by grillazz** [comment]
@tiangolo not sure what you mena by example app but here is code https://github.com/grillazz/fastapi-redis/blob/main/app/main.py where I added suggested monkey patch so if you just comment mokey patch lines it will be broken again

**Comment by pndiku** [comment]
> Now, for this specific problem, I would release a version pinning the version that worked if I was able to replicate the problem and be sure myself what's the error and what fixes it. But up to now, I haven't been able to replicate it once, and sadly no one has reported an example app that shows the problem. 

Try this: [http://idril.softedgeuganda.info:5000/docs](http://idril.softedgeuganda.info:5000/docs)

EDIT: When I use a VPN, it does seem to be a geolocated issue. Depending on my location, I get different views.

**Comment by nikokaoja** [comment]
Same here:
https://wps.neweuropeanwindatlas.eu/api/docs

though I've just fixed it with @ArcLightSlavik `swagger_monkey_patch`

**Comment by paymand** [comment]
Reported here: https://github.com/swagger-api/swagger-ui/issues/7290

**Comment by kpflugshaupt** [comment]
I'm glad to know this is neither on our end, nor on FastAPI's.
As of now, Switzerland still gets served the broken CSS/HTML/whatever it is, will check back tomorrow.

I for one would welcome this being delivered with FastAPI -- it's just the admin interface, I can live with it being a bit out of date.

**Comment by yshan2028** [comment]
同样遇到问题

**Comment by Hultner** [comment]
> Yeah, something around bundling a local version is on my plans, but those assets require an independent build step to include them, they can't/shouldn't be added to GitHub compiled. And it would require a new version of FastAPI every time they release a new version, so people would be stuck with old versions of Swagger and ReDoc even if new versions fixed bugs and added features, and it would increase the size of the distributable (but maybe that's not that significant?).

I would prefer the Swagger UI to be version pinned/fixed with the FastAPI version since this would give me a predictable reproducible artefact. For the same reason I want all my dependencies to be version locked before deploying.

I see it as beneficial if it works at build time it will also work whenever a user visits the page regardless if swagger have introduced a breaking change or an incompatible update (like this).

As for the particular problem I can no longer reproduce it.

**Comment by grillazz** [comment, marked as answer]
@tiangolo issue seems to be solved now.
